### DISCUSSION TITLE
How to complete code based on external source (e.g. , clangd / pyright /... )

### DISCUSSION THREAD

**Discussion by lainegates** [LazyVim/LazyVim#2344] [Q&A] [answered]
LazyVim is incredible tool that is fast, ui-friendly and out-of-box, which is the best vim environment I have tried.
But I still have some problem to switch to lazy-vim. The most important thing is code completion.
I have tried other famous vim plugins, such as `coc.nvim` / `space-vim`. They complete code with vim-plugins like `coc-clang` / `youcompleteme` for c++ language, or `pyright` for python language.

I checked plugins of lazy-vim, it seems that lazy-vim replies on tree-sitter to parse code. 
The problem I encountered is that  I am working on a huge project that complied with clang currently. Read the built output of clang ( i.e. use clangd) could speed up and give accurate completion information from the complicated project.

Thus, my question is how to let lazy-vim use clangd to complete code for c++ language ?
(Same for python language completion with pyright/... )

**Comment by trippel** [comment]
Did you try the lang extras?

https://www.lazyvim.org/extras/lang/clangd
https://www.lazyvim.org/extras/lang/python

If I am not mistaken, these should do what you are looking to achieve or am I missing something here?

**Reply by lainegates** [reply]
Yeah, I found it in https://www.lazyvim.org/extras/lang/clangd
Thank you very much!

**Comment by lainegates** [comment]
@trippel I read the clangd config in https://www.lazyvim.org/extras/lang/clangd.
It seems that lazy.nvim try to locate the root directory of clangd build directory.
```
 -- Correctly setup lspconfig for clangd 🚀
  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        -- Ensure mason installs the server
        clangd = {
          keys = {
            { "<leader>cR", "<cmd>ClangdSwitchSourceHeader<cr>", desc = "Switch Source/Header (C/C++)" },
          },
          root_dir = function(fname)
            return require("lspconfig.util").root_pattern(
              "Makefile",
              "configure.ac",
              "configure.in",
              "config.h.in",
              "meson.build",
              "meson_options.txt",
              "build.ninja"
            )(fname) or require("lspconfig.util").root_pattern("compile_commands.json", "compile_flags.txt")(
              fname
            ) or require("lspconfig.util").find_git_ancestor(fname)
          end,
          capabilities = {
            offsetEncoding = { "utf-16" },
          },
          cmd = {
            "clangd",
            "--background-index",
            "--clang-tidy",
            "--header-insertion=iwyu",
            "--completion-style=detailed",
            "--function-arg-placeholders",
            "--fallback-style=llvm",
          },
          init_options = {
            usePlaceholders = true,
            completeUnimported = true,
            clangdFileStatus = true,
          },
        },
      },
      setup = {
        clangd = function(_, opts)
          local clangd_ext_opts = require("lazyvim.util").opts("clangd_extensions.nvim")
          require("clangd_extensions").setup(vim.tbl_deep_extend("force", clangd_ext_opts or {}, { server = opts }))
          return false
        end,
      },
    },
  },
```

the clang build directory of my project is `/code/.clang-build`, the config setup clangd failed because the root directory is not found correctly.

If I add a clangd argument ` "--compile-commands-dir=/code/.clang-build",`, and the clangd setup normally. But this will change the lazy.nvim repo which is not I want.

I have some other arguments to set, could I set the clangd arguments at other place ? (e.g., in `~/.config/nvim/lua/config/lazy.lua` ? )  
I just read the lua grammar, and have no idea how to write the code. Is there some similar config example?

**Reply by trippel** [reply, marked as answer]
You could disable the extra and put it's content in a file in your plugin directory and configure it to your liking. You can overwrite parts of the config the same way without disabling the LazyExtra.
You will find inspiration in https://github.com/LazyVim/starter/blob/main/lua/plugins/example.lua.

**Reply by lainegates** [reply]
@trippel Thank you, I make it by following the example !
I edit it in `~/.config/nvim/lua/plugins/custom.lua`
```
  -- Correctly setup lspconfig for clangd 🚀
return  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        -- Ensure mason installs the server
        clangd = {
          -- overwrite default clangd settings
          cmd = {
            "clangd",
            "--background-index",
            "--clang-tidy",
            "--compile-commands-dir=/code/.clang-build",
            "--function-arg-placeholders",
            "--completion-style=detailed",
            "--fallback-style=Google",
            "--header-insertion=iwyu",
            "--j=4",
            "--malloc-trim",
            "--pch-storage=memory"
          },
          filetypes = { "c", "cpp", "objc", "objcpp", "cuda", "proto"},
        },
      },
    },
  }

```
